### ISSUE TITLE
Same seed yields different results

### ISSUE THREAD

**Issue by Quasimondo** [photogeniq/neural-texturize#5]
Running texturizer with the same seed results in slightly different results when repeated. I would expect the same seed to generated the identical texture every time. Here are three examples:

![crop1_gen](https://user-images.githubusercontent.com/182098/83141146-75128f80-a0ef-11ea-821b-029d1e199449.png)
![crop1_gen_a](https://user-images.githubusercontent.com/182098/83141156-79d74380-a0ef-11ea-9227-314bf5654fbf.png)
![crop1_gen_b](https://user-images.githubusercontent.com/182098/83141168-7e9bf780-a0ef-11ea-9249-f2a65fb51c23.png)

**Comment by Quasimondo** [comment]
Looks like this is caused by some non-deterministic CUDA functions - since when running the same seed on the CPU the results are identical.

From https://pytorch.org/docs/stable/notes/randomness.html:

> _There are some PyTorch functions that use CUDA functions that can be a source of nondeterminism. One class of such CUDA functions are atomic operations, in particular atomicAdd, which can lead to the order of additions being nondetermnistic. Because floating-point addition is not perfectly associative for floating-point operands, atomicAdd with floating-point operands can introduce different floating-point rounding errors on each evaluation, which introduces a source of nondeterministic variance (aka noise) in the result._

**Comment by htoyryla** [comment]
Try torch.backends.cudnn.deterministic = True ?

I use this all the time, appears to save memory.

**Comment by Quasimondo** [comment]
Ah, sorry I didn't mention that of course I had tried all the usual settings mentioned in the docs. It does not help.

**Comment by htoyryla** [comment]
Oh yes... good to know.

**Comment by alexjc** [comment]
It's good to know the CPU results are identical, that rules out issues bugs directly in this library ;-)

It's not clear to me what is causing the non-determinism on GPU.  'Conv2d` should become deterministic with `torch.backends.cudnn.deterministic = True`, but the other candidate in the encoder is only `torch.nn.AvgPool2d` — which has no warnings in the documentation about determinism.

It may be that the L-BFGS optimizer itself also introduces some non-determinism.  Testing a plain SGD optimizer would rule that out...

**Comment by Quasimondo** [comment]
I had first suspected the LBFGS solver and tentatively replaced it with Adam. The results look like crap, but they also differ when run on CUDA with the same seed.

I forgot to quote the most important line from that "Reprocibility" link above:
**

> "There is currently no simple way of avoiding nondeterminism in these functions."

**

**Comment by alexjc** [comment]
The interpolation would not be deterministic anyway, as soon as you change the content of the starting tensor, it would introduce different numbers and hence instability.
https://arxiv.org/abs/1705.02092v1

As for larger textures, I have a variety of code for tiling but not with gram matrices yet... I will see what's possible!
